File: arrays/3423-maximum-difference-between-adjacent-elements-in-a-circular-array.py
from typing import List
import unittest

# https://leetcode.com/problems/maximum-difference-between-adjacent-elements-in-a-circular-array/

# python3 -m unittest arrays/3423-maximum-difference-between-adjacent-elements-in-a-circular-array.py

class Solution(unittest.TestCase):
    def maxAdjacentDistance(self, nums: List[int]) -> int:
        return max(abs(nums[i]-nums[i-1]) for i in range(len(nums)))

    def test(self):
        for nums, expected in [
            ([1,2,4], 3),
            ([-5,-10,-5], 5),
        ]:
            output = self.maxAdjacentDistance(nums)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: arrays/3423-maximum-difference-between-adjacent-elements-in-a-circular-array.go
package arrays

// https://leetcode.com/problems/maximum-difference-between-adjacent-elements-in-a-circular-array/

func maxAdjacentDistance(nums []int) int {
	diff := -100
	n := len(nums)
	nei := n - 1
	for idx := 0; idx < n; idx++ {
		if nums[idx] > nums[nei] {
			diff = max(diff, nums[idx]-nums[nei])
		} else {
			diff = max(diff, nums[nei]-nums[idx])
		}
		nei = idx
	}
	return diff
}
